### PR TITLE
Adapt to upcoming Cabal installed package identifier format change

### LINKED ISSUE #851: Stack assumes IPIDs have a specific format
`parseGhcPkgId` assumes that an installed package ID is always a package name, a version, and a hash. IPIDs are opaque identifiers generated by Cabal, and Cabal reserves the right to change them (and will be changing them soon, c.f. https://github.com/haskell/cabal/issues/2745 )

### PR DESCRIPTION
**Pull request by snoyberg**
Fixes  #851

### CODE REVIEW

File: src/Stack/Types/GhcPkgId.hs [hunk 1]
```diff
@@ -36,13 +34,24 @@ instance Show GhcPkgIdParseFail where
...
 
 instance Hashable GhcPkgId
-instance Binary GhcPkgId
+instance Binary GhcPkgId where
+    put (GhcPkgId x) = do
+        -- magic string
+        putWord8 1
+        putWord8 3
+        putWord8 4
+        putWord8 7
```
Comment (chrisdone) [inline src/Stack/Types/GhcPkgId.hs]: What's this shenanigans?

**Comment by chrisdone** [comment]
I'm happy with this patch. Seems pretty mechanical. Laborious to do, but straight-forward updates.

File: src/Stack/Types/GhcPkgId.hs (hunk 1 above)
Comment (snoyberg) [inline src/Stack/Types/GhcPkgId.hs]: Make sure that the binary format being read is the same one that was written. See: https://github.com/commercialhaskell/stack/issues/889

**Comment by snoyberg** [comment]
I'd say: let's merge this in immediately after the next release is cut. I don't want to rock the boat on this release for such a non-urgent change, but would like to maximize testing during the next release cycle. Similar thoughts apply to #887. @borsboom sound good?

**Comment by chrisdone** [comment]
:+1:

**Pull request merged**